### ISSUE TITLE
 recordDeferredEvents() flushes the entire current queue, drops them completely recordEvents() fails twice

### ISSUE THREAD

**Issue by dorkusprime** [keen/keen-tracking.js#51]
recordDeferredEvents() utilizes the bulk tracking function, but flushes the entire queue – if the XHR fails twice in a row, these events are lost forever. This is not ideal for clients with potential connectivity issues, such as mobile devices.

Current implementation for reference:

``` javascript
function recordDeferredEvents(){
  var self = this, currentQueue;
  if (self.queue.capacity > 0) {
    currentQueue = JSON.parse(JSON.stringify(self.queue));
    self.queue = queue();
    self.queue.options = currentQueue.options;

    self.emit('recordDeferredEvents', currentQueue.events);
    self.recordEvents(currentQueue.events, function(err, res){
      if (err) {
        // Retry once
        self.recordEvents(currentQueue.events);
      }
      else {
        currentQueue = void 0;
      }
    });
  }
  return self;
}
```

A proposed solution would be to add the events _back into the queue_ upon 2x failures, rather than simply moving on.

**Comment by dorkusprime** [comment]
Note: This is closely related to https://github.com/keen/keen-tracking.js/issues/52

**Comment by Paytons-Epic** [comment]
It looks like a lot of headache is coming from the fact that a new queue is being created each time recordDeferredEvents() is called. If queue was replaced with a of long-lived spooler, then the errors and retry logic could be handled internally. This adds more complexity compared to just adding the events back into the new queue. I think the advantage is that you could have more control and clarity regarding the retry logic.

Are you open to a solution that replaces queue? I have a rough solution outlined and I could work on an implementation.

**Comment by dorkusprime** [comment]
Hi there!

So I think a spool implementation might very well be the best way to go here, but – you're right – it's certainly a more complex direction than the simple queue-replacement strategy going on right now, which means it'll will require a decent number of new tests to make sure it all works as intended.

I'll defer to @dustinlarimer on this, but it seems like a good overall direction to me.

**Comment by dorkusprime** [comment]
@Paytons-Epic Dustin's currently at a conference, so I got ahold of him – he says go for it! Just mind the testing :)

**Comment by Paytons-Epic** [comment]
Hey, thanks for getting right back to me. Also, I am excited that @dustinlarimer is interested to see more.

I did a little work on the spool this weekend and I am pretty happy with it (as a general concept). I had some trouble getting the environment up and running so I can't write any tests or build the final file yet. I am going to hop onto the developer group and see if I can get some help sorting out the dependency issues that I am seeing. The instructions at https://github.com/keen/keen-tracking.js/blob/master/CONTRIBUTING.md didn't work and I can't tell if it is just my computer (linux vs mac or wrong version of Node, npm, gulp, etc).

**Comment by dustinlarimer** [comment]
@Paytons-Epic sorry to hear about the install issues there - can you tell me more about your setup? Which versions of Node and NPM are you running? What errors (if any) are you seeing?

**Comment by Paytons-Epic** [comment]
@dustinlarimer I created a post in the developer group [here](https://groups.google.com/d/msg/keen-io-devs/fDFqohmgj2s/wQl--L07BgAJ). I tried to be very detailed, but let me know if you need more info. Hopefully I just overlooked something obvious and the solution is simple.

**Comment by dustinlarimer** [comment]
@Paytons-Epic I found a post that seems related - do you think these steps might help? https://github.com/Medium/phantomjs#im-on-debian-or-ubuntu-and-the-installer-failed-because-it-couldnt-find-node

> **I'm on Debian or Ubuntu, and the installer failed because it couldn't find node**
> Some Linux distros tried to rename node to nodejs due to a package conflict. This is a non-portable change, and we do not try to support this. The official documentation recommends that you run `apt-get install nodejs-legacy` to symlink `node` to `nodejs` on those platforms, or many NodeJS programs won't work properly.

**Comment by Paytons-Epic** [comment]
@dustinlarimer That is a good call. I have been messed up by aptitude many times in the past. I no longer trust them as a worthy source for dev tools. Just to be sure, I checked with "~$ which node" and "~$which nodejs". The former pointed to the correct install path and latter didn't find an install path. It looks like I avoided the above issue.

I also asked a friend with a mac to do a fresh install and no luck. It looks like he got the same error. I will keep searching to see if I can stumble onto an effective way to pin down the source of the error.

In the meantime, I will turn some attention toward understanding the current tests and trying to determine what new tests will be needed for the spooler.
